edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs

```diff
@@ -152,6 +152,22 @@ async Task ReconcileConnectionsAsync(HashSet<IIdentity> updatedIdentities)
 
                 await this.RemoveConnectionsAsync(identitiesRemoved);
                 await this.AddConnectionsAsync(identitiesAdded);
+
+                // When a device connects indirectly, then disconnects and connects directly again, the following happens:
+                // As we don't have "indirect disconnection", the device proxy instance lingers for a while before gets deleted
+                // by idle-time. Because of that it may show up in the "knownIdentities" list, and the way added connections are
+                // calculated above, it will not be added as a "direct connection" but keeps existing as "indirect connection".
+                // As a result, the device proxy will use mqtt topics for indirect connections which will not be subscribed
+                // by the directly connected device. As a solution, check if the broker sent connection-list and the known
+                // connections match by 'IsDirect' flag. If not, remove the indirect connection and recreate them as direct.
+                // Note, that a (child) $edgeHub is a direct connection but uses indirect dialect, so it has the IsDirect flag
+                // turned off, so ignore those cases from the list.
+                var reconnectedCandidates = new HashSet<IIdentity>(updatedIdentities);
+                var indirectConnections = this.knownConnections.Values.OfType<IDeviceProxy>()
+                                                                      .Where(p => !p.IsDirectClient && !(p.Identity is ModuleIdentity i && string.Equals(i.ModuleId, Constants.EdgeHubModuleId)))
```

Comment:
I know this is clear as is: `p => !p.IsDirectClient && !(p.Identity is ModuleIdentity i && string.Equals(i.ModuleId, Constants.EdgeHubModuleId))` but this seems like a common thing to filter on, so just curious if we already have a filter function that defines all this already that could be used?

Reply:
you mean something like "GetAllIndirectProxies()' or similar? There is a GetNestedConnectionsAsync() which returns the identities (and its used by another class from the outside). The 'this.knownConnections.Values.OfType<IDeviceProxy>().Where(p => !p.IsDirectClient.." part was unique until now, and 2 other cases have been added with this change, but all a bit different, so I am not sure what is the best way to move a common part to a method and if it is worth it.

Reply:
Yeah if it is unique no need to add anything extra :)